design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,363 @@
+
+## Domain Overview
+
+Magento allows customizing a product before adding the product to the cart.
+Customization happens through the customizable options defined by the merchant.
+Options can represent customization, other products, digital goods so on.
+
+All Magento product types represented through the options.
+
+
+### Definitions
+
+* **ProductOption** - represents a product characteristic which allows editing before adding to cart.
+ProductOption has to have a label, information on how option values should be displayed,
+and is this option mandatory.  
+* **ProductOptionValue**s - belong to a particular option as to a group and represent selections which allowed by the option.
+Option value could by display label, also one or may values could be pre-selected.  
+* **ProductVariant** - represent the final selection of one or multiple option values that will characterize the customized product in a shopping cart.
+Depends on the business scenario, a particular product variant could be linked with an existing product,
+with a price, or an inventory record, or no be linked to any.
+Even with no entities associated with the variant, it still has great value for a catalog because the presence of the variant says that such composition of options and their values described by the variant makes sense so that it can be purchased.
+   
+### Actual usages
+
+The application distinguishes two approaches to manage options:
+
+**Approach 1, "One to many"**: Multiple options selections lead a shopper to a selection of the single variation.
+
+Example: configurable products
+![](https://app.lucidchart.com/publicSegments/view/26cd0b67-13c1-44e4-8b61-cada36c67010/image.png)
+
+**Approach 2, "One to one"**: The selection of multiple options leads to multiple product variants.
+
+Examples: bundle product, customizable product, downloadable products.
+![](https://app.lucidchart.com/publicSegments/view/fea3f950-6f2f-4e46-90e4-f8ee4b6877f7/image.png)
+Both of approaches could be used together.
+
+Example: configurable product with customizable option.
+
+
+## Competitors analysis
+During my research, I analyzed documentation on how options are covered by the competitors.
+commercetools and Shopify attack this domain with product variants that represent all the possible options.
+Despite the domain simplification by reducing the number of entities both systems constrained with the size of a product which makes it impossible to compose complex business cases that we frequently may face in the Magento ecosystem.
```

Comment:
on commercetools a soft limitation of 100 is based on performance goals, they believe having product with more than 100 variants is not a good catalog design to begin with, same as having too many nested categories. In order to have good performance outcomes, they have applied this soft limit, which can be reviewed based on the project needs.

Reply:
Segregation of options and variants could resolve this limitation, in my opinion, all the limitations that we have at the moment as well as competitors, as for me, caused by exercise dependency on variants on PDP. If the variants matrix goes under the hood and will be used for filtering purposes mostly -  the most known performance limitations we know so far will be resolved.